docs/testing.md

```diff
@@ -0,0 +1,261 @@
+- **Avoid Docker dependency during development**: Developers can run `make test` without Docker installed
+- **Enable targeted CI workflows**: CI can run unit tests quickly on every commit and integration tests less frequently
+
+Without the `-tags=integration` flag, Go will skip all integration test files.
+
+## Test Organization
+
+### Unit Tests
+- Located alongside source files (e.g., `internal/app/app_test.go`)
+- No build tags required
+- Should be fast (< 100ms per test)
+- Mock external dependencies
+
+### Integration Tests
+- Located in `integration/` package
+- Require `//go:build integration` tag
+- Use real Docker services via `internal/testutil/` harness
+- Test complete workflows end-to-end
+
+## Writing Tests
+
+### Writing a Unit Test
+
+Create a `*_test.go` file in the same package:
+
+```go
+package app_test
+
+import (
+    "testing"
+    "github.com/simone-viozzi/bosun/internal/app"
+)
+
+func TestFeature(t *testing.T) {
+    // Arrange
+    a := app.New()
+
+    // Act
+    result := a.DoSomething()
+
+    // Assert
+    if result != expected {
+        t.Errorf("got %v, want %v", result, expected)
+    }
+}
+```
+
+### Writing an Integration Test
+
+Create a test file in `integration/` with the build tag:
+
+```go
+//go:build integration
+// +build integration
+
+package integration
+
+import (
+    "context"
+    "testing"
+    "time"
+    "github.com/simone-viozzi/bosun/internal/testutil"
+)
+
+func Test_Integration_MyFeature(t *testing.T) {
+    t.Parallel()
+
+    ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+    defer cancel()
+
+    // Start Docker Compose stack
+    stack := testutil.StartCompose(t, ctx, "docker-compose.yaml")
+
+    // Get service port
+    port := testutil.HostPort(t, ctx, stack.Project, "service-name", 80)
+
+    // Test your feature...
+}
+```
+
+## Test Utilities
+
+The `internal/testutil/` package provides helpers for integration tests:
+
+- **`StartCompose(t, ctx, files...)`**: Starts a Docker Compose stack with automatic cleanup
+- **`HostPort(t, ctx, project, service, containerPort)`**: Gets the published host port for a service
+- **`DumpLogs(t, ctx, project, outDir)`**: Dumps container logs to a directory (useful for debugging failures)
+
+Each test gets a unique Docker Compose project name to enable parallel execution.
+
+## Troubleshooting
+
+### Docker Not Running
+
+**Error:** `Cannot connect to the Docker daemon`
+
+**Solution:** Start Docker:
+```bash
+# On Linux with systemd
+sudo systemctl start docker
+
+# On macOS/Windows
+# Start Docker Desktop application
+```
+
+### Port Collisions
+
+**Error:** `Bind for 0.0.0.0:XXXX failed: port is already allocated`
+
+**Solution:** The test harness assigns random host ports to avoid collisions. If you see this error:
+1. Check for other services using the port: `lsof -i :XXXX` (macOS/Linux)
+2. Ensure the compose file doesn't specify fixed host ports (use `"80"` not `"8080:80"`)
```

Comment:
same here, this is not specific to this repo, remove or point to the official documentation